Updated Stencil Test Case to prevent it from disable after every draw

Vaivaswat2244 opened the pull request:
<!--
  Thank you for contributing! Please use this pull request (PR) template.


 In the description field of this PR, include "resolves #XXXX" tagging the issue you are fixing. If this PR addresses the issue but doesn't completely resolve it (ie the issue should remain open after your PR is merged), write "addresses #XXXX".-->
Resolves #7554 

 Changes:
<!-- Add here what changes were made in this pull request and if possible provide links showcasing the changes. -->
- Introduced this.isStencilTestOn and this._userEnabledStencil to track stencil state.
- Overrode WebGL's enable(), disable(), and getEnabled() methods to ensure:
   - The stencil test is marked as user-enabled only when explicitly set outside clip().
   - pop() restores the correct stencil state after clipping operations.
- Modified stencil test logic:
   - Prevent unnecessary disable(STENCIL_TEST) calls when it is user-enabled.
   - Restore user-enabled stencil test state after pop().

 Screenshots of the change:

 Used this snippet in preview/index.html,
 ```javascript

<body>
  <script type="module">
    import p5 from '../src/app.js';

    const sketch = function (p) {
      let gl;

      p.setup = function () {
        p.createCanvas(400, 400, p.WEBGL);
        p.noStroke();
        gl = p._renderer.GL;
        gl.enable(gl.STENCIL_TEST);
      };

      p.draw = function () {
        p.background(0);
        gl.clear(gl.STENCIL_BUFFER_BIT);

        // Set stencil value to 1
        gl.stencilFunc(gl.ALWAYS, 1, ~0);
        gl.stencilOp(gl.KEEP, gl.REPLACE, gl.REPLACE);

        p.fill(255, 0, 0);
        p.plane(200, 400);

        // Only draw where stencil value is 1
        gl.stencilFunc(gl.EQUAL, 1, ~0);
        gl.stencilOp(gl.KEEP, gl.KEEP, gl.KEEP);

        p.fill(0, 0, 255);
        p.plane(400, 200);

        p.noLoop();
      };
    };

    new p5(sketch);
  </script>
</body>
</html>

 ```
SCREENSHOT: 
![Screenshot from 2025-02-26 19-30-52](https://github.com/user-attachments/assets/08a9a98b-ee4b-442f-adcd-da08abc8dbac)

<!-- If applicable, add screenshots depicting the changes. -->

#### PR Checklist
<!--
  To check any option, replace the "[ ]" with a "[x]". Be sure to check out how it looks in the Preview tab! Feel free to remove any portion of the template that is not relevant for your issue.
-->

- [x] `npm run lint` passes
- [ ] [Inline reference] is included / updated
- [ ] [Unit tests] are included / updated

[Inline reference]: https://p5js.org/contribute/contributing_to_the_p5js_reference/
[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

davepagurek commented on src/webgl/p5.RendererGL.js:
```diff
@@ -427,6 +427,46 @@ class RendererGL extends Renderer {
+            // When pop() disables the stencil test after clip(), 
+            // restore the user's stencil test setting
+            if (this._clipDepth === this._pushPopDepth) {
+                this.isStencilTestOn = this._userEnabledStencil;
```
It's a little unintuitive that calling `disable()` elsewhere in the code might not actually disable. How do you feel about making `enable`/`disable` only record the last value so that `getEnabled` is fast, and then place this logic of what value to revert to elsewhere?

The idea would be, write the code so that the behaviour would be the same even if we haven't overridden `enable` and `disable`: e.g. when we begin clipping, we can record `this.isStencilTestOn = this.drawingContext.getEnabled(this.drawingContext.STENCIL_TEST)`, and then in `pop` where we turn off stencil testing currently, we would either turn it on or off based on the value of `this.isStencilTestOn`.

Then, this overriding of `enable`/`disable` just makes the `getEnabled` call run faster, without changing what it does or its side effects.

davepagurek commented:
Thanks for taking this task on, it's looking good! I have some comments on how we might rearrange the implementation to make the code easier for other contributors to read.

davepagurek commented:
One other thought: if stencil testing was on, after we finish clipping in `pop`, do we want to clear the stencil buffer? otherwise it effectively just continues clipping.

davepagurek reviewed (commented):
### Some thoughts on the overall logic

Just looking again at the if statements in the `drawingContext.enable` override: Is the if statement in here trying to detect if the user is manually calling `drawingContext.disable` vs if p5 itself is doing it?

One idea to try to simplify this: Maybe we should take those `prevEnable` and `prevDisable` variable we have, and store it on the renderer with names like `internalEnable` and `internalDisable`. Calling those just enable/disable; and then we're free to make our overrides a lot simpler, just:
```js
this.drawingContext.enable = (key) => {
  if (key === this.drawingContext.STENCIL_TEST) {
    this._userEnabledStencil = true
  }
  return prevEnable.call(key)
```

Then:
- wherever we internally call `GL.enable(GL.STENCIL_TEST)`, we'd call `internalEnable(GL.STENCIL_TEST)` instead so that it keeps `_userEnabledStencil` the same
- in all the places where we currently just disable stencil testing, we'd only actually disable it if `_userEnabledStencil` is false too, and otherwise leave it on.

This way we also don't need to record what the value is right before we enable stencil testing, since we don't really need to know what it was right before if we've recorded what the user has set it to.

I think that would mean we don't need the `isStencilTestOn` check as well, and might simplify some of our logic. Do you think that would work?

### Clearing the buffer on `pop`

Thinking a little more about the use case here, clearing the stencil buffer in `pop` would only make a difference if the user already manually enabled the stencil buffer, and additionally is using p5 clipping. I think that's actually not a combination we need to worry about, since the stencil buffer does clipping, if you're turning it on manually, the user won't be using p5's clipping. So maybe we take the simpler approach and don't do anything additional here.

### Testing

Sorry for all the back-and-forth on this, it turns out the logic is complicated haha. That's maybe a good sign that we should make some unit tests involving manual calls to enable/disable the stencil test?

Vaivaswat2244 commented:
> ### Some thoughts on the overall logic
> Just looking again at the if statements in the `drawingContext.enable` override: Is the if statement in here trying to detect if the user is manually calling `drawingContext.disable` vs if p5 itself is doing it?
> 
> One idea to try to simplify this: Maybe we should take those `prevEnable` and `prevDisable` variable we have, and store it on the renderer with names like `internalEnable` and `internalDisable`. Calling those just enable/disable; and then we're free to make our overrides a lot simpler, just:
> 
> ```js
> this.drawingContext.enable = (key) => {
>   if (key === this.drawingContext.STENCIL_TEST) {
>     this._userEnabledStencil = true
>   }
>   return prevEnable.call(key)
> ```
> 
> Then:
> 
> * wherever we internally call `GL.enable(GL.STENCIL_TEST)`, we'd call `internalEnable(GL.STENCIL_TEST)` instead so that it keeps `_userEnabledStencil` the same
> * in all the places where we currently just disable stencil testing, we'd only actually disable it if `_userEnabledStencil` is false too, and otherwise leave it on.
> 
> This way we also don't need to record what the value is right before we enable stencil testing, since we don't really need to know what it was right before if we've recorded what the user has set it to.
> 
> I think that would mean we don't need the `isStencilTestOn` check as well, and might simplify some of our logic. Do you think that would work?
> 
> ### Clearing the buffer on `pop`
> Thinking a little more about the use case here, clearing the stencil buffer in `pop` would only make a difference if the user already manually enabled the stencil buffer, and additionally is using p5 clipping. I think that's actually not a combination we need to worry about, since the stencil buffer does clipping, if you're turning it on manually, the user won't be using p5's clipping. So maybe we take the simpler approach and don't do anything additional here.
> 
> ### Testing
> Sorry for all the back-and-forth on this, it turns out the logic is complicated haha. That's maybe a good sign that we should make some unit tests involving manual calls to enable/disable the stencil test?

Sure, got it.
I have made the changes required. Tests are not included rn, once you confirm these I can write tests involving manual calls to enable/disable the stencil test.

davepagurek commented on src/webgl/p5.RendererGL.js:
```diff
@@ -1387,7 +1423,7 @@ class RendererGL extends Renderer {
 
     const gl = this.GL;
-    gl.clearStencil(0);
+    gl.clearStencil(0); 
```
I think there's a trailing space at the end here that we can take out.

Vaivaswat2244 replied:
Sure

davepagurek commented on src/webgl/p5.RendererGL.js:
```diff
@@ -427,6 +427,39 @@ class RendererGL extends Renderer {
+      if (key === this.drawingContext.STENCIL_TEST) {
+        // When pop() disables the stencil test after clip(),
+        // preserve the user's stencil test setting
+        if (this._clipDepth === this._pushPopDepth) {
```
Do we need this condition here, since we're calling `_internalDisable` in `pop`? I think maybe this is only called now when users manually call `disable`, so we can set `this._userEnabledStencil = true`. The same may be true in `enable` as well if we update the spots where we turn on stencil testing to use `_internalEnable`

Vaivaswat2244 replied:
Right! I'll do this for enable as well.

davepagurek commented on src/webgl/p5.RendererGL.js:
```diff
@@ -427,6 +427,39 @@ class RendererGL extends Renderer {
     this.scratchMat3 = new Matrix(3);
+
+    this._userEnabledStencil = false;
+    this.isStencilTestOn = false;
```
Do we use this any more?

Vaivaswat2244 replied:
Right, we don't.

davepagurek reviewed (commented):
The updates are looking good! I think the general approach looks almost complete. There are a few cases being handled that I think we may not need to handle any more and can be omitted. But in general this looks almost done, feel free to start adding tests!

Vaivaswat2244 commented:
So, I have added the tests for stencil testing which I thought were appropriate. Let me know to add or remove some cases. Also replaced the internalEnable for enable in the file.

davepagurek commented on test/unit/webgl/p5.RendererGL.js:
```diff
@@ -2782,5 +2883,5 @@ suite('p5.RendererGL', function() {
       myp5.checkPMatrix();
     });
-  });
+  });  
```
Looks like some trailing whitespace got added here, can we take that out?

Vaivaswat2244 replied:
Right!

davepagurek commented on test/unit/webgl/p5.RendererGL.js:
```diff
@@ -2704,6 +2705,106 @@ suite('p5.RendererGL', function() {
+        myp5.clip(() => {
+          myp5.rect(0, 0, 10, 10);
+        });
+        myp5.pop();
```
Maybe right before the `pop()` we can `assert.equal(gl.isEnabled(gl.STENCIL_TEST), true)` to double-check that it does get turned on correctly within the clipping block? (Possibly also in the test case above too)

Vaivaswat2244 replied:
Okay, sure

davepagurek reviewed (commented):
Sorry for the delay, I think this looks great! Nice work on the tests! I left two very minor comments, overall I think this is just about ready to go!

Vaivaswat2244 commented:
Made the changes. This should be good to go now ig
